include/tscpp/api/HttpMsgComp.h

```diff
@@ -0,0 +1,449 @@
+destructors of these classes.  Therefore, blocks containing instances of these claesses must end before the call to
+TSHttpTxnReenable().
+*/
+
+namespace atscppapi
+{
+class MsgBase;
+
+class MimeField
+{
+public:
+  // If loc is TS_NULL_MLOC, instance constructed in empty state.
+  //
+  explicit MimeField(MsgBase &msg, TSMLoc loc = TS_NULL_MLOC) : _msg(msg), _loc(loc) {}
+
+  // MimeField at (zero-base) index idx in HTTP message.
+  //
+  MimeField(MsgBase &msg, int idx);
+
+  // MimeField with given name in HTTP message.
+  //
+  MimeField(MsgBase &msg, std::string_view name);
+
+  TSMLoc
+  loc() const
+  {
+    return _loc;
+  }
+
+  MsgBase &
+  msg() const
+  {
+    return _msg;
+  }
+
+  // Valid means non-empty.
+  //
+  bool valid() const;
+
+  // Put instance into empty state, releasing resources as appropriate.
+  //
+  void reset();
+
+  // Can not copy, only move.
+
+  MimeField(const MimeField &) = delete;
+  MimeField &operator=(const MimeField &) = delete;
+
+  MimeField(MimeField &&source);
+
+  // Move assign only allowed if the two instances refer to the same message.
+  //
+  MimeField &operator=(MimeField &&source);
+
+  void destroy();
+
+  // Next field, returns invalid instance if none.
+  //
+  MimeField next() const;
+
+  // Next field with same name, return invalid instance if none.
+  //
+  MimeField nextDup() const;
+
+  std::string_view nameGet() const;
+
+  void nameSet(std::string_view);
+
+  void valuesClear();
+
+  // Returns number of values. Values index are from 0 to valuesCount() - 1.
+  //
+  int valuesCount() const;
+
+  std::string_view valGet(int idx) const;
+
+  void valSet(int idx, std::string_view new_value);
+
+  // Insert a new value at index idx.  All values with index >= idx prior to calling this have their index incremented by one.
+  //
+  void valInsert(int idx, std::string_view new_value);
+
+  // Append a new value at the end.
+  //
+  void valAppend(std::string_view new_value);
+
+  ~MimeField();
+
+private:
+  MsgBase &_msg;
+  TSMLoc _loc;
+};
+
+class MsgBuffer : noncopyable
+{
+public:
+  MsgBuffer(TSMBuffer hndl) : _hndl(hndl) {}
+
+  // Returns the handle.
+  //
+  TSMBuffer
+  bufHndl() const
+  {
+    return _hndl;
+  }
+
+protected:
+  MsgBuffer() {}
+
+  void
+  set(TSMBuffer hndl)
+  {
+    _hndl = hndl;
+  }
+
+private:
+  TSMBuffer _hndl;
+};
+
+class MsgBase : public MsgBuffer
+{
+public:
+  MsgBase(TSMBuffer bHndl, TSMLoc msgLoc) : MsgBuffer(bHndl), _msgLoc(msgLoc) {}
+
+  TSMLoc
+  msgLoc() const
+  {
+    return _msgLoc;
+  }
+
+  bool
+  hasMsg() const
+  {
+    return msgLoc() != TS_NULL_MLOC;
+  }
+
+  //  Returns number of mime header lines in HTTP message.  Can only be called if hasMsg() is true.
+  //
+  int
+  mimeFieldsCount() const
+  {
+    assert(hasMsg());
+
+    return TSMimeHdrFieldsCount(bufHndl(), msgLoc());
+  }
+
+  MimeField
+  newMimeField()
+  {
+    assert(hasMsg());
```

Comment:
Why don't use the `ink_assert` or `ink_release_assert` ?

Reply:
This header is exported to plugin code, so it has to be TSAssert().  See line 31, I removed the use of C assert.  I verified that  __OPTIMIZE__ is defined by both clang and gcc.  The problem with using TSAssert (or ink_assert) directly is that (as in the check above) you also don't want any functions in the boolean expression called in the non-debug build.  (Although, it this particular case, it wouldn't matter as hasMsg() would be inlined and elided by the optimizer.)